### DISCUSSION TITLE
Openid configuration (dex)

### DISCUSSION THREAD

**Discussion by Bender01-code** [DependencyTrack/dependency-track#1219] [Q&A]
I use DEX and different secrets (CLIENT_SECRET) for auth in applications.

I did not find any "CLIENT_SECRET" config in dependency-track. 

Do you have any chance to add the option in a future release?

**Comment by jedrzej-andrykowski** [comment]
I also have problem to configure Dependency Track with google accounts where `client_secret` is provided. When I try to authorize through OpenId I got an error:

```
{
  "error": "invalid_request",
  "error_description": "client_secret is missing."
}
```

As @Bender01-code wrote: there is no `client_secret` config, so there is no possibility to achieve this? Maybe there is some workaround?

**Reply by prasadasokan** [reply]
> I also have problem to configure Dependency Track with google accounts where `client_secret` is provided. When I try to authorize through OpenId I got an error:
> 
> ```
> {
>   "error": "invalid_request",
>   "error_description": "client_secret is missing."
> }
> ```
> 
> As @Bender01-code wrote: there is no `client_secret` config, so there is no possibility to achieve this? Maybe there is some workaround?

I got this working with the following ENV variables added. 

**apiserver:**

```
      - ALPINE_OIDC_ENABLED=true
      - ALPINE_OIDC_ISSUER=https://accounts.google.com
      - ALPINE_OIDC_USERNAME_CLAIM=email
      - ALPINE_OIDC_USER_PROVISIONING=true
      - ALPINE_OIDC_CLIENT_ID=xxxxxxxxxx-************************.apps.googleusercontent.com
```
**frontend:**

```
      - OIDC_ISSUER=https://accounts.google.com
      - OIDC_CLIENT_ID=xxxxxxxxxx-************************.apps.googleusercontent.com
      - OIDC_SCOPE=profile openid email
      - OIDC_FLOW=implicit
      - OIDC_LOGIN_BUTTON_TEXT=Google SSO
```

**Comment by nscuro** [comment]
The OIDC authentication is performed by DT's frontend, in the user's browser. Not by the API server or "backend". In [OIDC lingo](https://oauth.net/2/client-types/), the client is *public*, not *confidential*. Client secrets don't make any sense for public clients.

Dependency-Track can be configured to use either the [Code+PKCE](https://oauth.net/2/pkce/) (default) or [Implicit](https://oauth.net/2/grant-types/implicit/) OIDC flow. I'm not familiar with Google as IdP or DEX, but maybe you can use the implicit flow which doesn't require a client secret.

**Reply by GHSebamo** [reply]
I am wanting to use Dependency-track at work, and my company's policy is to for our Keycloak realms to be "Confidential" rather than "Public".
Am I right to understand that I can't configure OIDC because of this ?

**Reply by nscuro** [reply]
That is correct. Confidential clients are not supported. A client secret doesn't make sense here, as it would be immediately disclosed. 

I'd be interested in why your security team has that policy, because it means that you cannot use OIDC with SPAs or mobile apps, at all.

**Reply by GHSebamo** [reply]
Something about their being a gateway between the user and Keycloak that generates a key that allows the user to access a specific application. So with this architecture, a key received for an application can't be used to login to another application thanks to the secret key shared by an application and Keycloak (using "Confidential'). But a key obtained when connecting to an application that uses a Public Flow would allow someone to log into any of our other applications with that same key.

I'm sorry that this probably makes no sense, I'm trying to pass on their answer, but it's not my specialty, it wasn't a written answer and I'm trying to translate it from another language.

PS : my answer is for intranet applications, not internet.

**Comment by boulc** [comment]
I just encountered the same issue and found out we can just specify `public: true` instead of a secret in DEX configuration. The PKCE flow completes successfully with the settings below.

Source: PKCE implementation https://github.com/dexidp/dex/pull/1784

apiserver:
```
    - ALPINE_OIDC_ENABLED=true
    - ALPINE_OIDC_ISSUER=https://dex.********************/dex
    - APLINE_OIDC_CLIENT_ID=dependency-track
    - ALPINE_OIDC_USERNAME_CLAIM=preferred_username
    - ALPINE_OIDC_TEAMS_CLAIM=groups
    - ALPINE_OIDC_USER_PROVISIONING=true
    - ALPINE_OIDC_TEAM_SYNCHRONIZATION=true 
 ```

frontend:
```
      - OIDC_ISSUER=https://dex.********************/dex
      - OIDC_CLIENT_ID=dependency-track
      - OIDC_SCOPE=openid profile email groups
      - OIDC_FLOW=code
      - OIDC_LOGIN_BUTTON_TEXT=DEX
```

dex:
```
staticClients:
  - id: dependency-track
    redirectURIs:
      - 'https://dependencytrack.********************/static/oidc-callback.html'
    name: 'Dependency-Track'
    public: true
```
